plugin.rb

```diff
@@ -308,33 +361,54 @@ def self.absolute(raw)
       url
     end
 
-    def self.set_filter_by_id(id, channel, filter, channel_id = nil)
-      data = get_store(id)
+    def self.set_filter(channel, filter, category_id = nil, tags = nil)
+      tags = [] if tags.nil?
+      id = SecureRandom.hex(16)
 
-      update = data.index {|i| i['channel'] === channel || i['channel'] === channel_id }
+      data = { channel: channel, filter: filter, category_id: category_id, tags: tags }
 
-      if update
-        data[update]['filter'] = filter
-        data[update]['channel'] = channel # fix old IDs
-      else
-        data.push({ channel: channel, filter: filter })
-      end
+      ::PluginStore.set(PLUGIN_NAME, "filter_#{id}", data)
+
+      ::PluginStore.set(PLUGIN_NAME, "_category_#{category_id}_#{channel}", id) if category_id.present?
+      tags.each{ |t| ::PluginStore.set(PLUGIN_NAME, "_tag_#{t}_#{channel}", id) }
+      return id
+    end
+
+    def self.update_filter(id, channel, filter, category_id = nil, tags = [])
+      data = get_filter(id)
 
-      data = data.uniq { |i| i['channel'] }
+      old_channel = data['channel']
+      old_category_id = data['channel']
+      old_tags = data['tags']
 
-      ::PluginStore.set(PLUGIN_NAME, "category_#{id}", data.uniq)
+      data['filter'] = filter
+      data['channel'] = channel # fix old IDs
+      data['category_id'] = category_id
+      data['tags'] = tags
+
+      ::PluginStore.set(PLUGIN_NAME, "filter_#{id}", data)
+
+      ::PluginStore.remove(PLUGIN_NAME, "_category_#{old_category_id}_#{old_channel}") if old_category_id.present?
+      old_tags.each{|t| ::PluginStore.remove(PLUGIN_NAME, "_tag_#{t}_#{old_channel}")}
+      ::PluginStore.set(PLUGIN_NAME, "_category_#{category_id}_#{channel}", id) if category_id.present?
+      tags.each{|t| ::PluginStore.set(PLUGIN_NAME, "_tag_#{t}_#{channel}", id)}
     end
 
-    def self.delete_filter(id, channel)
-      data = get_store(id)
-      data.delete_if do |i|
-        i['channel'] === channel
-      end
-      ::PluginStore.set(PLUGIN_NAME, "category_#{id}", data)
+    def self.delete_filter(id)
+      ::PluginStore.remove(PLUGIN_NAME, "filter_#{id}")
+      ::PluginStoreRow.where(plugin_name: PLUGIN_NAME, value: id).where("key ~* :pat", :pat => '^_.*').destroy_all
```

Comment:
Actually I used it existing slack plugin code [#L54](https://github.com/discourse/discourse-slack-official/blob/master/plugin.rb#L54). Here it is fetching all records which starting with the char "_".